usd/src/sdf_parser/Joint.cc

```diff
@@ -0,0 +1,402 @@
+/*
+ * Copyright (C) 2022 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include "sdf/usd/sdf_parser/Joint.hh"
+
+#include <ignition/math/Angle.hh>
+#include <ignition/math/Pose3.hh>
+#include <ignition/math/Vector3.hh>
+
+// TODO(ahcorde) this is to remove deprecated "warnings" in usd, these warnings
+// are reported using #pragma message so normal diagnostic flags cannot remove
+// them. This workaround requires this block to be used whenever usd is
+// included.
+#pragma push_macro ("__DEPRECATED")
+#undef __DEPRECATED
+#include <pxr/base/gf/quatf.h>
+#include <pxr/base/gf/vec3f.h>
+#include <pxr/base/tf/token.h>
+#include <pxr/usd/sdf/path.h>
+#include <pxr/usd/usd/relationship.h>
+#include <pxr/usd/usd/stage.h>
+#include <pxr/usd/usdPhysics/driveAPI.h>
+#include <pxr/usd/usdPhysics/fixedJoint.h>
+#include <pxr/usd/usdPhysics/joint.h>
+#include <pxr/usd/usdPhysics/prismaticJoint.h>
+#include <pxr/usd/usdPhysics/revoluteJoint.h>
+#include <pxr/usd/usdPhysics/sphericalJoint.h>
+#pragma pop_macro ("__DEPRECATED")
+
+#include "sdf/Error.hh"
+#include "sdf/Joint.hh"
+#include "sdf/JointAxis.hh"
+#include "sdf/Link.hh"
+#include "sdf/Model.hh"
+#include "../UsdUtils.hh"
+
+namespace sdf
+{
+// Inline bracke to help doxygen filtering.
+inline namespace SDF_VERSION_NAMESPACE {
+//
+namespace usd
+{
+  /// \brief Helper function for setting a USD joint's pose relative to the
+  /// joint's parent and child.
+  /// \param[in] _jointPrim The USD joint prim
+  /// \param[in] _joint The SDF representation of _jointPrim
+  /// \param[in] _parentModel The SDF model that is the parent of _joint
+  /// \return UsdErrors, which is a vector of UsdError objects. Each UsdError
+  /// includes an error code and message. An empty vector indicates no error
+  /// occurred when setting _joint's pose relative to _joint's parent and child
+  UsdErrors SetUSDJointPose(pxr::UsdPhysicsJoint &_jointPrim,
+      const sdf::Joint &_joint, const sdf::Model &_parentModel)
+  {
+    UsdErrors errors;
+
+    ignition::math::Pose3d parentToJoint;
+    if (_joint.ParentLinkName() == "world")
+    {
+      ignition::math::Pose3d modelToJoint;
+      auto poseErrors = usd::PoseWrtParent(_joint, modelToJoint);
+      if (!poseErrors.empty())
+        return poseErrors;
+
+      // it is assumed the _parentModel's parent is the world
```

Comment:
The SDF -> USD converter does not support nested models, which is mentioned in #845. `sdf::usd::ParseSdfModel` also exits early if a nested model is attempted to be parsed: https://github.com/ignitionrobotics/sdformat/blob/5059bb8b8dd3e946bf84c0654ed073d32278e931/usd/src/sdf_parser/Model.cc#L53-L59

However, I guess that downstream users could still try to call the `sdf::usd::ParseSdfJoint` method directly and have unexpected behavior if they're using a nested model. What about I update the documentation of `sdf::usd::ParseSdfJoint` to mention that the model that `_parentModel`'s parent must be the world, and cannot be a nested model?

Reply:
I think you just update this comment to say: "it is assumed the _parentModel's parent is the world because nested models are not yet supported (see issue #845)"